programs/test/selftest.c

```diff
@@ -304,6 +305,65 @@ int main( int argc, char *argv[] )
         mbedtls_exit( MBEDTLS_EXIT_FAILURE );
     }
 
+    /*
+     * The C standard allows padding bits in the representation
+     * of standard integer types, but our code does currently not
+     * support them.
+     *
+     * Here we check that the underlying C implementation doesn't
+     * use padding bits, and fail cleanly if it does.
+     *
+     * The check works by casting the maximum value representable
+     * by a given integer type into the unpadded integer type of the
+     * same bit-width and checking that it agrees with the maximum value
+     * of that unpadded type. For example, for a 4-byte int,
+     * MAX_INT should be 0x7fffffff in int32_t. This assumes that
+     * CHAR_BIT == 8, which is checked in check_config.h.
+     */
+
+#define CHECK_PADDING_SIGNED(TYPE, NAME)                                \
+    do                                                                  \
+    {                                                                   \
+        if( ( sizeof( TYPE ) == 2 &&                                    \
+              (int16_t) NAME ## _MAX != 0x7FFF )             ||         \
+            ( sizeof( TYPE ) == 4 &&                                    \
+              (int32_t) NAME ## _MAX != 0x7FFFFFFF )         ||         \
+            ( sizeof( TYPE ) == 8 &&                                    \
+              (int64_t) NAME ## _MAX != 0x7FFFFFFFFFFFFFFF ) )          \
```

Comment:
Currently some library modules work on platforms without a 64-bit type. It may be useful to keep `selftest.c` working in this case, even if we don't officially support those platforms (primarily because we don't intend to make all modules work without a 64-bit type).

Reply:
@gilles-peskine-arm How could we detect that?

Reply:
`UINTMAX_T` if available. Or `defined(ULLONG_MAX)`.

Reply:
This bloats the code significantly. Given that this is only the selftest and we're not claiming to work without 64-bit types, I'd prefer to leave things as they are, unless you insist on the change.

Reply:
I have not addressed this one - I think the conclusion here is that it's not worth the complexity?

Reply:
I think getting the completely portable test to work is too much work. As for platforms without a 64-bit type, back then we targeted C89, but now we target C99, where a 64-bit type is officially mandatory. We don't insist on every feature of C99, but a 64-bit type was commonly available even on C89 platforms, so I think it's reasonable to insist that it's present now. Thus, I no longer think a change is needed here.